agent/lib/kontena/workers/service_pod_worker.rb

```diff
@@ -126,16 +128,21 @@ def current_state
       end
     end
 
-    # @param [String] current_state
-    def sync_state_to_master(current_state)
-      data = {
+    # @param current_state [String]
+    # @param error [Exception]
+    def sync_state_to_master(current_state, error = nil)
+      state = {
         service_id: service_pod.service_id,
         instance_number: service_pod.instance_number,
+        rev: service_pod.deploy_rev,
         state: current_state,
-        rev: service_pod.deploy_rev
+        error: error ? "#{error.class}: #{error}" : nil,
       }
-      rpc_client.async.notification('/node_service_pods/set_state', [node.id, data])
-      @prev_state = current_state
+
+      if state != @prev_state
+        rpc_client.async.notification('/node_service_pods/set_state', [node.id, state])
```

Comment:
Because this is an async notification, we won't notice if the RPC gets dropped, and the server is left waiting... if this was changed to a `rpc_client.request` instead, we would get an error if the RPC fails... then we could crash and retry sooner? OTOH, I suppose right now we'll eventually re-send this notification, so it should still make progress ATM.